LibSWBF2.NET/Wrappers/Level.cs

```diff
@@ -74,14 +75,12 @@ public bool IsWorldLevel
         public Model[] GetModels()
         {
             APIWrapper.Level_GetModels(NativeInstance, out IntPtr modelArr, out uint modelCount);
-            IntPtr[] modelPtrs= new IntPtr[modelCount];
-            Marshal.Copy(modelArr, modelPtrs, 0, (int)modelCount);
-
-            Model[] models = new Model[modelCount];
-            for (int i = 0; i < modelCount; i++)
+            
+            Model[] models = MemUtils.IntPtrsToWrappers<Model>(modelArr, (int) modelCount);
+            
+            foreach (Model model in models)
             {
-                models[i] = new Model(modelPtrs[i]);
-                Children.Add(new WeakReference<NativeWrapper>(models[i]));
+                Children.Add(new WeakReference<NativeWrapper>(model));                
```

Comment:
Not exactly sure what these refs are for?  I get that a Level should hold children, but why are they ```WeakReferences``` ?

Reply:
Because imagine this scenario:
```
Level someLevel = Level.FromFile("some.lvl");
Model someModel = someLevel.GetModel("someModelInsideLVL");

// set's assume that someLevel was the only reference to that level instance
someLevel = null;

// somewhere else in the code
if (someModel != null)
    // someModel wont be null at this point, but printing the Name will fail
    Console.WriteLine(someModel.Name);
```

The c++ `Level` holds all the c++ Model instances, so it's responsible for c++ Model creation and deletion.
When setting C# `someLevel` to `null`, the GC will collect the C# `Level` instance at some point. When this happens, C# `~Level()` will be called, which will then delete the c++ Level instance, which results in all c++ Model pointers fetched from that c++ `Level` to become invalid.
But now, thanks to the _magic_ of VM languages, the C# `someModel` instance is still alive (we cannot force the language to set all C# `Model` references to `null`) and still holds it's `IntPtr NativeInstance` member pointing to a c++ `Model`, which no longer exists, so the call `someModel.Name` will just crash.
The whole purpose of the `Childrens` list in C# `Level` is to keep references to all C# `Model` instances ever given out via `GetModels()` or `GetModel(string modelName)`, so the moment the C# `Level `instance gets deleted (and so the c++ `Level`) we _invalidate_ all C# `Model` instances (via resetting their `IntPtr NativeInstance` to `null`) belonging to that C# `Level`. Storing WeakReferences instead of actual ones has the advantage of not prohibiting the GC to collect a certain C# `Model` instance before the whole C# `Level` get's collected.